core/verify_manager.go

```diff
@@ -0,0 +1,134 @@
+}
+
+func NewVerifyManager(blockchain *BlockChain) *VerifyManager {
+	verifiedCache, _ := lru.New(verifiedCacheSize)
+	vm := &VerifyManager{
+		bc:            blockchain,
+		tasks:         make(map[common.Hash]*VerifyTask),
+		verifiedCache: verifiedCache,
+		verifyCh:      make(chan common.Hash),
+		exitCh:        make(chan struct{}),
+	}
+	return vm
+}
+
+func (vm *VerifyManager) Start() {
+	//read disk store to initial verified cache
+	//load unverified blocks in a normalized chain and start a batch of verify task
+	header := vm.bc.CurrentHeader()
+	vm.NewBlockVerifyTask(header)
+	go vm.mainLoop()
+}
+
+func (vm *VerifyManager) Stop() {
+	defer close(vm.exitCh)
+	vm.exitCh <- struct{}{}
+}
+
+func (vm *VerifyManager) mainLoop() {
+	prune := time.NewTicker(time.Second)
+	defer prune.Stop()
+	for {
+		select {
+		case hash := <-vm.verifyCh:
+			vm.cacheBlockVerified(hash)
+			rawdb.WriteTrustBlockHash(vm.bc.db, hash)
+			delete(vm.tasks, hash)
+		case <-prune.C:
+			for hash, task := range vm.tasks {
+				if vm.bc.CurrentHeader().Number.Uint64()-task.blockHeader.Number.Uint64() > 15 {
+					task.terminalCh <- struct{}{}
+					delete(vm.tasks, hash)
+				}
+			}
+		case <-vm.exitCh:
+			return
+		}
+	}
+}
+
+func (vm *VerifyManager) NewBlockVerifyTask(header *types.Header) {
+	for i := 0; i < 11; i++ {
+		hash := header.Hash()
+		diffLayer := vm.bc.GetTrustedDiffLayer(hash)
+		//if this block has no diff, there is no need to verify it.
+		if diffLayer == nil {
+			continue
+		}
+		//if verified cache record that this block has been verified, skip.
+		if _, ok := vm.verifiedCache.Get(hash); ok {
+			continue
+		}
+		//if verified storage record that this block has been verified, skip.
+		if rawdb.ReadTrustBlockHash(vm.bc.db, hash) {
+			vm.cacheBlockVerified(hash)
+			continue
+		}
+		//if there already has a verify task for this block, skip.
+		if _, ok := vm.tasks[hash]; ok {
+			continue
+		}
+		diffHash, err := GetTrustedDiffHash(diffLayer)
+		if err != nil {
+			log.Error("failed to get diff hash", "block", hash, "number", header.Number, "error:", err)
+		}
+		verifyTask := NewVerifyTask(diffHash, header, vm.peers, vm.bc.db, vm.verifyCh, vm.allowUntrustedVerify)
+		vm.tasks[hash] = verifyTask
+		header = vm.bc.GetHeaderByHash(header.ParentHash)
+	}
+}
+
+func (vm *VerifyManager) cacheBlockVerified(hash common.Hash) {
+	if vm.verifiedCache.Len() >= verifiedCacheSize {
+		vm.verifiedCache.RemoveOldest()
+	}
+	vm.verifiedCache.Add(hash, true)
+}
+
+//CheckAncestorVerified function check whether head 11 of this block has been verified.
+//If not, the blockchain should stop to insert new block.
+func (vm *VerifyManager) CheckAncestorVerified(header *types.Header) bool {
+	pHeader := header
+	for i := 0; i < 11; i++ {
+		pHeader = vm.bc.GetHeaderByHash(pHeader.ParentHash)
+	}
+	hash := pHeader.Hash()
+	if _, ok := vm.verifiedCache.Get(hash); ok {
+		return true
+	} else {
+		return rawdb.ReadTrustBlockHash(vm.bc.db, hash)
+	}
+	return false
```

Comment:
Won't get here, you can delete else, return rawdb.ReadTrustBlockHash(vm.bc.db, hash) directly.

Reply:
> Won't get here, you can delete else, return rawdb.ReadTrustBlockHash(vm.bc.db, hash) directly.

done